--- /tmp/Z8001MB/cpm8k/biosboot.s	2026-02-07 16:52:17.000000000 -0500
+++ bios/biosboot.s	2026-02-07 16:39:18.000000000 -0500
@@ -55,7 +55,7 @@
 ! *
 ! ****************************************************
 
-	.equ	SYSTEM, 0x03000000		! system address
+	.equ	SYSTEM, 0x0B000000		! system address
 	.equ	SYSSTK, (SYSTEM + 0x0BFFE)	! system stack top
 
 ! ****************************************************
--- /tmp/Z8001MB/cpm8k/biosdef.s	2026-02-07 16:52:17.000000000 -0500
+++ bios/biosdef.s	2026-02-07 16:39:18.000000000 -0500
@@ -10,7 +10,7 @@
 ! ******************************************************************************
 !	if ID_SPLIT = 1, I and D space split supported		
 ! ******************************************************************************	
-	.equ	ID_SPLIT, 1
+	.equ ID_SPLIT, 1
 
 ! ******************************************************************************
 ! *
@@ -134,4 +134,3 @@
 	.equ	PSECSZ, 512
 
 
-
--- /tmp/Z8001MB/cpm8k/biosif.s	2026-02-07 16:52:17.000000000 -0500
+++ bios/biosif.s	2026-02-07 16:39:18.000000000 -0500
@@ -94,9 +94,9 @@
 	push	@r15, #_wboot	! set up system stack so that a return will warm boot
 	
 	call	trapinit	! set up traps, then enable interrupts
-	ei	vi, nvi
+	!ei	vi, nvi
 
-	call	scc_init	! set up serial port and prrint a message
+	!call	scc_init	! set up serial port and prrint a message
 	lda	r4, bootmsg
 	call	puts
 
--- /tmp/Z8001MB/cpm8k/biostrap.s	2026-02-07 16:52:17.000000000 -0500
+++ bios/biostrap.s	2026-02-07 16:39:18.000000000 -0500
@@ -273,9 +273,10 @@
 	NONSEG
 	ret
 
-
-!fp_epu:
-!        ret
+/*
+fp_epu:
+        ret
+*/
 /*
 dbgbc:  
 	! Segmented mode
--- /tmp/Z8001MB/cpm8k/bioscall4.s	2026-02-07 16:52:17.000000000 -0500
+++ bios/bioscall4.s	2026-02-07 16:39:18.000000000 -0500
@@ -72,7 +72,8 @@
 	
 memtbl:
 	.if ID_SPLIT == 1
-	.word	0x05
+	!.word	0x05
+	.word	0x04
 	.long	0x0A000000		! Region 1
 	.long	0x00010000		!  merged I and D
 	.long	0x08000000		! Region 2
@@ -81,8 +82,8 @@
 	.long	0x00010000		!  separeted D
 	.long	0x0B000000		! Region 4
 	.long	0x00010000		!  accessing I as D
-	.long	0x09000000		! for DDT?
-	.long	0x00010000	
+	!.long	0x09000000		! for DDT?
+	!.long	0x00010000	
 	
 	.else
 	.word	0x04
--- /tmp/Z8001MB/cpm8k/diskio.s	2026-02-07 16:52:17.000000000 -0500
+++ bios/diskio.s	2026-02-07 16:39:18.000000000 -0500
@@ -3,20 +3,22 @@
 !   Disk I/O subroutines 
 !
 !   Copyright(c) 2020 4sun5bu
-
+!
+!
 	.global	diskrd, diskwr, disk_init
+	
+	.equ	IDE_DATAW, 0x01c0	! Word access
+	.equ	IDE_DATAB, 0x01c1	! Byte access
+	.equ	IDE_ERROR, 0x01c3
+	.equ	IDE_FEATURES, 0x01c3
+	.equ	IDE_SECT_CNT, 0x01c5
+	.equ	IDE_SECT_NUM, 0x01c7
+	.equ	IDE_CYL_LOW, 0x01c9
+	.equ	IDE_CYL_HIGH, 0x01cb
+	.equ	IDE_DEV_HEAD, 0x01cd
+	.equ	IDE_STATUS, 0x01cf
+	.equ	IDE_COMND, 0x01cf
 
-	.equ	IDE_DATAW, 0x0020	! Word access
-	.equ	IDE_DATAB, 0x0021	! Byte access
-	.equ	IDE_ERROR, 0x0023
-	.equ	IDE_FEATURES, 0x0023
-	.equ	IDE_SECT_CNT, 0x0025
-	.equ	IDE_SECT_NUM, 0x0027
-	.equ	IDE_CYL_LOW, 0x0029
-	.equ	IDE_CYL_HIGH, 0x002b
-	.equ	IDE_DEV_HEAD, 0x002d
-	.equ	IDE_STATUS, 0x002f
-	.equ	IDE_COMND, 0x002f
 
 	.equ	BSY_BIT, 7
 	.equ	DRDY_BIT, 6
@@ -46,15 +48,41 @@
 
 	unsegm
 	sect	.text
+	.macro	CMDRDY
+
+cmdrdy_loop\@:
+	inb	rl0, #IDE_STATUS
+	bitb	rl0, #BSY_BIT
+	jr	nz, cmdrdy_loop\@
+	bitb	rl0, #DRDY_BIT
+	jr	z, cmdrdy_loop\@
+	.endm
+
 
+	.macro	DATRDY
+dat_loop\@:
+	inb	rl0, #IDE_STATUS
+	bitb	rl0, #BSY_BIT
+	jr	nz, dat_loop\@
+	bitb	rl0, #DRQ_BIT
+	jr	z, dat_loop\@
+	.endm
 !------------------------------------------------------------------------------
 disk_init:
 	CHKBSY
 	CHKDRDY
-	ldb	rl0, #0x01		! Set to byte access 
+	
+	ldb	rl0, #0x01		! Byte access 
 	outb	#IDE_FEATURES, rl0
-	ldb	rl0, #0xef		! 
+	ldb	rl0, #0xef
 	outb	#IDE_COMND, rl0
+	CHKBSY
+
+	ldb	rl0, #0x82
+	outb	#IDE_FEATURES, rl0
+	ldb	rl0, #0xef
+	outb	#IDE_COMND, rl0
+	CHKBSY
 	ret 
 	
 !------------------------------------------------------------------------------
@@ -64,8 +92,9 @@
 !	    r4  --- Buffer Address
 
 diskrd:
+	CMDRDY
 	CHKBSY
-	CHKDRDY
+	
 	ldb	rl0, #1
 	outb	#IDE_SECT_CNT, rl0
 	outb	#IDE_SECT_NUM, rl3
@@ -76,16 +105,16 @@
 	outb	#IDE_DEV_HEAD, rh2
 	ldb	rl0, #0x20		! data in command
 	outb	#IDE_COMND, rl0
-	CHKDRQ
-	CHKBSY
+	DATRDY
+
 	inb	rl0, #IDE_STATUS	! Reset INTRQ
+	ld	r2, #512
 1:
 	inb	rl0, #IDE_DATAB
 	ldb	@r4, rl0
 	inc	r4, #1
-	inb	rl0, #IDE_STATUS    ! Check if data exists 
-	bitb	rl0, #DRQ_BIT
-	jr 	nz, 1b
+	djnz    r2,1b
+
 	inb	rl0, #IDE_STATUS    ! 
 	ret
 	
@@ -98,6 +127,7 @@
 diskwr:
 	CHKBSY
 	CHKDRDY
+
 	ldb	rl0, #1
 	outb	#IDE_SECT_CNT, rl0
 	xorb	rl0, rl0
@@ -109,16 +139,16 @@
 	outb	#IDE_DEV_HEAD, rh2
 	ldb	rl0, #0x30		! data out command
 	outb	#IDE_COMND, rl0
-	CHKDRQ
-	CHKBSY
+	DATRDY
+
 	inb	rl0, #IDE_STATUS	! Reset INTRQ
+	ld	r2, #512
 1:
 	ldb	rl0, @r4
 	outb	#IDE_DATAB, rl0
 	inc	r4, #1
-	inb	rl0, #IDE_STATUS	! Check if data exists 
-	bitb	rl0, #DRQ_BIT
-	jr	nz, 1b
+	djnz    r2,1b
+	
 	CHKBSY
 	inb	rl0, #IDE_STATUS
 	ret
--- /tmp/Z8001MB/cpm8k/z8530.s	2026-02-07 16:52:17.000000000 -0500
+++ bios/z8530.s	2026-02-07 16:39:18.000000000 -0500
@@ -4,8 +4,8 @@
 !    
 !   Copyright(c) 2020 4sun5bu 
 
-    .equ SCCAC, 0x0005
-    .equ SCCAD, 0x0007
+    .equ	SCCAC, 0x00C3
+    .equ	SCCAD, 0x00C1
 
     .global  scc_init, scc_out, scc_in, scc_status
 
@@ -67,18 +67,13 @@
 !------------------------------------------------------------------------------
     sect .rdata
 
+
 scccmds:
-    !.byte   9, 0xc0     ! Reset
-    .byte   3, 0xc0     ! Receiver disable
-    .byte   5, 0xe2     ! Transmiter disable
-    .byte   4, 0x44     ! x16, 1stop-bit, non-parity
-    .byte   3, 0xe0     ! Receive  8bit/char 
-    .byte   5, 0xe2     ! Send 8bit/char dtr rts
-    .byte   11, 0x50    ! BG use for receiver and transmiter
-    .byte   12, 37      ! 4800bps at 6MHz clock
-    .byte   13, 00
-    .byte   14, 0x02    ! PCLK for BG
-    .byte   14, 0x03    ! BG enable
-    .byte   3, 0xe1     ! Receiver enable
-    .byte   5, 0xea     ! Transmiter enable
+	.byte	0x18		! Reset 0x18
+	.byte	4, 0xc4		! x64, 1stop-bit, non-parity (7Mhz/64=115200)
+	.byte	1, 0x04		! Disable Interrupts
+	.byte	3, 0xe1		! Receive  8bit/char, rts auto         
+	.byte	5, 0xea		! Send 8bit/char, dtr rts
+	.byte	0xc0		! Reset OVERRUN
 scccmde:
+
--- /tmp/Z8001MB/cpm8k/cpmsys.x	2026-02-07 16:52:17.000000000 -0500
+++ bios/cpmsys.x	2026-02-07 16:39:18.000000000 -0500
@@ -3,8 +3,8 @@
 ENTRY(_start)
 
 MEMORY
-{
-	CPMSYS : ORIGIN = 0x30000, LENGTH = 64k
+{ 
+	CPMSYS : ORIGIN = 0xb0000, LENGTH = 64k
 	/* ROM : ORIGIN =    , LENGTH = */ 
 }
 
--- /dev/null	2026-02-07 16:59:24.000000000 -0500
+++ bios/i8251.s	2026-02-07 16:39:18.000000000 -0500
@@ -0,0 +1,103 @@
+!------------------------------------------------------------------------------
+! i8251.s
+!  Serial I/O routines for SCC
+!
+!  Copyright (c) 2022 tpaxia
+!------------------------------------------------------------------------------
+
+	 .global  scc_init, scc_out, scc_in, scc_status
+	
+	sect	.text
+	unsegm
+
+
+	.equ	c8253ctrl, 0x0127   ! control register 
+	.equ	c8253c0, 0x0121		! counter 0 (TTY/printer timing)
+	.equ    c8253c1, 0x0123		! counter 1 (keyboard timing)
+
+	.equ	TTYC, 0xC3	
+	.equ	TTYD, 0xC1
+
+!------------------------------------------------------------------------------
+! initscc
+!   Initialize i8353 Timer for 19200 baud rate
+!
+!   destroyed:  r2, r3, r4 
+
+scc_init:
+	! set up TTY divisor
+
+	ldb  rl2, #0x36
+	outb #c8253ctrl,rl2
+	ld  r2, #0x0080
+	outb #c8253c0,rl2
+	outb #c8253c0,rh2
+
+	! set up Keyboard divisor
+
+	ldb  rl2, #0x76
+	outb #c8253ctrl,rl2
+	ld  r2, #0x0080
+	outb #c8253c1,rl2
+	outb #c8253c1,rh2
+
+	ld      r2, #(scccmde2 - scccmds2)    ! initialize 8251
+	ld      r3, #TTYC
+	ld      r4, #scccmds2
+	otirb   @r3, @r4, r2
+	ret
+	
+	 
+
+!------------------------------------------------------------------------------
+! scc_out 
+!   output 1 byte to the serial port
+!   input:      r5 --- Ascii code
+!   destroyed:  rl0
+
+scc_out:
+	inb	rl0, #TTYC
+	andb	rl0, #0x01
+	jr	z, scc_out
+	outb	#TTYD, rl5
+	ret
+
+!------------------------------------------------------------------------------
+! scc_in
+!   input 1 byte from the serial port
+!   return:     r7 --- read data
+!   destroyed:  r0, r1
+
+scc_in:
+	inb	rl0, #TTYC
+	andb	rl0, #0x02
+	jr	z, scc_in
+	clr     r7
+	inb	rl7, #TTYD
+	ret
+!------------------------------------------------------------------------------
+! scc_status
+!   return:     r7 --- 0xff:data exists, 0x00:no data
+!   destroyed:  r1, r10
+
+scc_status:
+    inb     rl0, #TTYC
+    clr     r7
+    andb    rl0, #0x02
+    ret     z
+    com     r7
+    ret
+
+
+!------------------------------------------------------------------------------
+	sect .rodata
+
+scccmds2:
+	.byte	0x0		! Sync mode 
+	.byte	0x0		! Sync char1 
+	.byte	0x0		! Sync char2
+	.byte	0x40	! Reset device
+	.byte	0x4d	! ASync mode, 1x, N, 8, 1
+	.byte	0x37	! Err reset, enable TX & RX
+	.byte	0x27	! Err reset, enable TX & RX
+scccmde2:
